presto-main/src/main/java/io/prestosql/metadata/TableHandle.java

```diff
@@ -17,23 +17,31 @@
 import com.fasterxml.jackson.annotation.JsonProperty;
 import io.prestosql.connector.ConnectorId;
 import io.prestosql.spi.connector.ConnectorTableHandle;
+import io.prestosql.spi.connector.ConnectorTableLayoutHandle;
+import io.prestosql.spi.connector.ConnectorTransactionHandle;
 
-import java.util.Objects;
+import java.util.Optional;
 
 import static java.util.Objects.requireNonNull;
 
 public final class TableHandle
 {
     private final ConnectorId connectorId;
     private final ConnectorTableHandle connectorHandle;
+    private final ConnectorTransactionHandle transaction;
+    private final Optional<ConnectorTableLayoutHandle> layout;
```

Comment:
To me it looks like to much of information to be stored in `TableHandle`. To me `TableHandle` represents an handle to a table, same way like a column handle. It is expected to be lightweight. What you have here is designed for very specific use case, hence maybe a new entity should be created for that.

Notice that you are planning to put even more information here (aggregations, joins, projections), the more you add here the less "table"-like it becomes, instead it becomes a more generic "relation" (no longer a table). Also notice that `TableHandle` is used for insert (perfect fit usage) and now with your changes it becomes questionable.

Reply:
> To me it looks like to much of information to be stored in TableHandle. To me TableHandle represents an handle to a table, same way like a column handle. It is expected to be lightweight. What you have here is designed for very specific use case, hence maybe a new entity should be created for that.

Storing the TableLayout is only temporary. I plan to eventually get rid of them entirely, but didn't do so yet because it requires backward incompatible changes to the SPI. 

But at the end of the day, yes, the plan is for TableHandle to potentially represent more than raw tables. The object doesn't need to be heavyweight -- it's up to the connector to decide what it holds on to internally (it could be just an id to an stored definition, etc). As to whether it becomes more "relation"-like, that's not much of a concern. In fact, the SQL spec treats everything as some form of a table:

```
A table is a collection of zero or more rows where each row is a sequence of one or more column values. 

[...] 

A table is either a base table, a derived table, or a transient table.

[...]

A derived table is a table derived directly or indirectly from one or more other tables by the evaluation of an expression, 
such as a <joined table>, <data change delta table>, <query expression>, or <table expression>. A <query expression> 
can contain an optional <order by clause>
```

> Also notice that TableHandle is used for insert (perfect fit usage) and now with your changes it becomes questionable.

That's ok, because whether a table can be inserted into is decided at analysis time. So you'd never end up with a TableHandle to insert into that represents something that can't be inserted into. BTW, did you know the SQL spec allows for inserting into things that are more complex that simple tables? For example, there's an entire section describing "updatable views".

Reply:
I see your point. Anyway, I still think that `TableHandle` was a simple wrapper around `ConnectorTableHandle` and there are still places where it is used as such. Now, they will get a complex type, so they would need to at least check if the layout is empty there.

> Storing the TableLayout is only temporary. I plan to eventually get rid of them entirely, but didn't do so yet because it requires backward incompatible changes to the SPI.

But there will be something else that will replace `TableLayout`. So `TableHandle` won't become any simpler.

> So you'd never end up with a TableHandle to insert into that represents something that can't be inserted into.

Having separate a type would make expressive in the code that such case is impossible (without knowing the actual flow).

> BTW, did you know the SQL spec allows for inserting into things that are more complex that simple tables?

Mind blowing. Anyway, I don't think that `TableHandle` that is suited for read will match that.